streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -519,6 +520,10 @@ void commit(final boolean startNewTransaction, final Map<TopicPartition, Long> p
             }
         } catch (final CommitFailedException | ProducerFencedException error) {
             throw new TaskMigratedException(this, error);
+        } catch (final RetriableCommitFailedException error) {
```

Comment:
@ableegoldman This is the only place I think should be augmented: I checked all the callers of commitSync / commitAsync and it seems this is the only place that needs to be updated.

The reason we can safely ignore it is the following: 1) in streams we would not commit if streams instance knows it is in the middle of a rebalance, so 2) the only case we will hit this is if the generation was reset and a rebalance is underway, which should be very rare, PLUS in this case EOS is disabled so ignoring and crashing would just cause duplicated records, which is fine.

Reply:
Not sure I understand why we can be sure eos is disabled if we hit this case -- also, does that mean if we do try to commit during a rebalance with eos enabled we will still get a fatal exception rather than the retriable one?

Reply:
If EOS is enabled, we do not call consumer.commit but with producer.sendOffsetsInTxn; in the latter case there's no special handling logic inside producer for `REBALANCING_IN_PROGRESS` so it would be treated as:

```
fatalError(new KafkaException("Unexpected error in TxnOffsetCommitResponse: " + error.message()));
```

Reply:
Ok. But so to be extra clear, we should still always block commits from happening during a rebalance when eos is used, yes? 
I'm guessing it has to be that way since that makes sense, but if you have any thoughts about a possible alternative fix for eos let's at least make a ticket for now. I guess so long as we can still return data during a rebalance then 429 is still useful for active throughput of eos..

Reply:
> Ok. But so to be extra clear, we should still always block commits from happening during a rebalance when eos is used, yes?

That's right.

For EOS, the producer.sendOffsetsToTxn would not see a `REBALANCE_IN_PROGRESS` error since it is sending a "blind offset update" to the group coordinator which does not check its generation or member-id at all, since the producer is already be guarded by the txn coordinator by its transactional-id.

Reply:
Right, because the txn coordinator is not the same as the group coordinator. So then, do you as the user just need to be very careful not to commit with the producer while the consumer on the other end is currently going through a rebalance?
It would be nice to find a real solution to this besides just saying "eos can't ever try to commit", as some point in the (possibly distant) future -- although I suppose this could actually end up resolved as a side-effect of KIP-447? @mjsax @abbccdda @cpettitt-confluent (tagging you all because I'm not sure who's on that KIP anymore)

Reply:
Are you asking about the behavior of Kafka Streams specifically or a general pattern where users use a combo of consumer and producer to achieve EOS?

Assuming you were asking about the latter case, in that scenario the consumer-producer combo is not fenced by the group coordinator any more but only by the txn coordinator, and like I said, the producer.sendOffsetsToTxn is a blind offset commit to group coordinator which is not going to check generation (I'm talking about prior KIP-447 here) -- and hence they have to follow the similar pattern as current Streams that a single partition fetched by the consumer is only mapped to a single producer transactional-id so that if that partition moves to a different consumer, the corresponding producer transactional-id has to be changed accordingly otherwise there's no fencing.

With KIP-447, the API would be augmented to:

```
sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets, ConsumerGroupMetadata consumerGroupMetadata)
```

And what I'm thinking then is that ConsumerGroupMetadata can contain the `state` of the consumer as well, and callers should NOT send offsets to transaction if the state is in REBALANCING.

Does that make sense?

Reply:
Yes! Thanks for the full explanation. I guess I overlooked that any general user of consumer+producer+eos would actually have to follow that same pattern, which we use in Streams -- obvious in hindsight, because why else would we do it.